readthedocs/doc_builder/environments.py

```diff
@@ -267,7 +270,88 @@ def get_wrapped_command(self):
                                    for part in self.command]))))
 
 
-class BuildEnvironment(object):
+class BaseEnvironment(object):
+
+    """
+    Base environment class.
+
+    Used to run arbitrary commands outside a build.
+    """
+
+    def __init__(self, project, environment=None):
+        # TODO: maybe we can remove this Project dependency also
+        self.project = project
+        self.environment = environment or {}
+        self.commands = []
+
+    def record_command(self, command):
+        pass
+
+    def _log_warn_only(self, msg):
+        log.warn(LOG_TEMPLATE.format(
+            project=self.project.slug,
+            version='latest',
+            msg=msg,
+        ))
+
+    def run(self, *cmd, **kwargs):
+        """Shortcut to run command from environment."""
+        return self.run_command_class(cls=self.command_class, cmd=cmd, **kwargs)
+
+    def run_command_class(self, cls, cmd, warn_only=False, **kwargs):
+        """
+        Run command from this environment.
+
+        Use ``cls`` to instantiate a command
+
+        :param warn_only: Don't raise an exception on command failure
+        """
+        # Remove PATH from env, and set it to bin_path if it isn't passed in
+        env_path = self.environment.pop('BIN_PATH', None)
+        if 'bin_path' not in kwargs and env_path:
+            kwargs['bin_path'] = env_path
+        assert 'environment' not in kwargs, "environment can't be passed in via commands."
+        kwargs['environment'] = self.environment
+
+        # TODO: ``build_env`` is passed as ``kwargs`` when it's called from a ``*BuildEnvironment``
+        # kwargs['build_env'] = self
+
+        build_cmd = cls(cmd, **kwargs)
+        self.commands.append(build_cmd)
+        build_cmd.run()
+
+        # TODO: I don't like how it's handled this entry point here
+        if not warn_only:
+            # TODO: maybe receive ``record`` as an attribute for skip/record
+            # just specific commands but not all of them ran under the
+            # *BuildEnvironment
```

Comment:
Hehe.

Yes, I was thinking on having the `record` attribute passed in the `run` method to override the one from the `__init__` from `*BuildEnvironment` classes (now we have `LocalEnvironment` and `(Local/Docker)BuildEnvironment`)

Reply:
Yeah, i think a `record` arg on `run()` makes sense. Given `__init__(record=False, ...)`, and `run(record=True, ...)`, what do think we should do? I'd probably say don't record if `__init__(record=False)`, thoughts?

Reply:
I think it's better to follow what `run(record=)` is set independently how was the class initialized.

Why? Because sometimes the `LocalEnvironment` is initialized far away (in another file and just passed as argument) from the code where the command is ran and you want/need to override the default record.

I implemented this at https://github.com/rtfd/readthedocs.org/blob/e1766a4d1061f7e97c75039b84a196088b111164/readthedocs/doc_builder/environments.py#L310-L312